packages/core-ethereum/crates/core-ethereum-misc/src/redeem.rs

```diff
@@ -0,0 +1,669 @@
+    db: Arc<RwLock<Db>>,
+    channel: &ChannelEntry,
+    onchain_tx_sender: &impl Fn(AcknowledgedTicket) -> F,
+) -> Result<usize>
+where
+    Db: HoprCoreEthereumDbActions,
+    F: futures::Future<Output = std::result::Result<String, String>>,
+{
+    let channel_id = channel.get_id();
+
+    let count_redeemable_tickets = db
+        .read()
+        .await
+        .get_acknowledged_tickets(Some(*channel))
+        .await?
+        .iter()
+        .filter(|t| t.status == Untouched)
+        .count();
+    info!("there are {count_redeemable_tickets} acknowledged tickets in channel {channel_id} which can be redeemed");
+
+    // Return fast if there are no redeemable tickets
+    if count_redeemable_tickets == 0 {
+        return Ok(0);
+    }
+
+    // Keep holding the DB write lock until we mark all the eligible tickets as BeginRedeemed
+    let mut to_redeem = Vec::new();
+    {
+        // Lock the database and retrieve again all the redeemable tickets
+        let mut db = db.write().await;
+        let redeemable = db
+            .get_acknowledged_tickets(Some(*channel))
+            .await?
+            .into_iter()
+            .filter(|t| Untouched == t.status);
+
+        for mut avail_to_redeem in redeemable {
+            if let Err(e) = set_being_redeemed(db.deref_mut(), &mut avail_to_redeem, *EMPTY_TX_HASH).await {
+                error!("failed to update state of {}: {e}", avail_to_redeem.ticket)
+            } else {
+                to_redeem.push(avail_to_redeem);
+            }
+        }
+    }
+
+    info!(
+        "{} acknowledged tickets are still available to redeem in {channel_id}",
+        to_redeem.len()
+    );
+
+    let redeemed = futures::future::join_all(to_redeem.into_iter().map(|ack_ticket| async {
+        let ticket_id = ack_ticket.to_string();
+        if let Err(e) = unchecked_ticket_redeem(db.clone(), ack_ticket, onchain_tx_sender).await {
+            error!("failed to redeem {ticket_id}: {e}");
+            false
+        } else {
+            true
+        }
+    }))
+    .await
+    .into_iter()
+    .filter(|r| *r)
+    .count();
+
+    info!("{redeemed} tickets have been redeemed in channel {channel_id}");
+    Ok(redeemed)
+}
+
+async fn unchecked_ticket_redeem<Db, F>(
+    db: Arc<RwLock<Db>>,
+    mut ack_ticket: AcknowledgedTicket,
+    on_chain_tx_sender: impl Fn(AcknowledgedTicket) -> F,
+) -> Result<Hash>
+where
+    Db: HoprCoreEthereumDbActions,
+    F: futures::Future<Output = std::result::Result<String, String>>,
+{
+    set_being_redeemed(db.write().await.deref_mut(), &mut ack_ticket, *EMPTY_TX_HASH).await?;
+    debug!("sending {} for on-chain redemption", ack_ticket.ticket);
+    match (on_chain_tx_sender)(ack_ticket.clone()).await {
+        Ok(tx_hash_str) => {
+            let tx_hash = Hash::from_hex(&tx_hash_str)?;
+            debug!("{ack_ticket} tx {tx_hash} completed");
+
+            db.write().await.mark_redeemed(&ack_ticket).await?;
```

Comment:
We mark the ticket as redeemed without on-chain confirmation. Is that intended? The tx could still fail on-chain at this point.

Reply:
The function actually awaits the on-chain confirmation.